stackslib/src/net/unsolicited.rs

```diff
@@ -0,0 +1,1030 @@
+            &self.local_peer,
+            outbound_neighbor_key_opt.or_else(|| { self.check_peer_authenticated(event_id) }),
+            &new_microblocks.index_anchor_block,
+            new_microblocks.microblocks.len()
+        );
+
+        // do we have the associated anchored block?
+        match chainstate.get_block_header_hashes(&new_microblocks.index_anchor_block) {
+            Ok(Some(_)) => {
+                // yup; can process now
+                debug!("{:?}: have microblock parent anchored block {}, so can process its microblocks", &self.local_peer, &new_microblocks.index_anchor_block);
+                !buffer
+            }
+            Ok(None) => {
+                if buffer {
+                    debug!(
+                        "{:?}: Will buffer unsolicited MicroblocksData({})",
+                        &self.local_peer, &new_microblocks.index_anchor_block
+                    );
+                    true
+                } else {
+                    debug!(
+                        "{:?}: Will not buffer unsolicited MicroblocksData({})",
+                        &self.local_peer, &new_microblocks.index_anchor_block
+                    );
+                    false
+                }
+            }
+            Err(e) => {
+                warn!(
+                    "{:?}: Failed to get header hashes for {:?}: {:?}",
+                    &self.local_peer, &new_microblocks.index_anchor_block, &e
+                );
+                false
+            }
+        }
+    }
+
+    /// Determine if an unsolicited NakamotoBlockData message contains data we can potentially
+    /// buffer
+    pub(crate) fn is_nakamoto_block_bufferable(
+        &mut self,
+        sortdb: &SortitionDB,
+        chainstate: &StacksChainState,
+        nakamoto_block: &NakamotoBlock,
+    ) -> bool {
+        if chainstate
+            .nakamoto_blocks_db()
+            .has_nakamoto_block(&nakamoto_block.block_id())
+            .unwrap_or(false)
+        {
+            debug!(
+                "{:?}: Aleady have Nakamoto block {}",
+                &self.local_peer,
+                &nakamoto_block.block_id()
+            );
+            return false;
+        }
+
+        let mut can_process = true;
+        let sn = match SortitionDB::get_block_snapshot_consensus(
+            &sortdb.conn(),
+            &nakamoto_block.header.consensus_hash,
+        ) {
+            Ok(Some(sn)) => sn,
+            Ok(None) => {
+                debug!(
+                    "No sortition {} for block {}",
+                    &nakamoto_block.header.consensus_hash,
+                    &nakamoto_block.block_id()
+                );
+                // we don't have the sortition for this, so we can't process it yet (i.e. we need
+                // to buffer)
+                can_process = false;
+                // load the tip so we can load the current reward set data
+                self.burnchain_tip.clone()
```

Comment:
Is this always the correct snapshot to query for the reward set?

For example, let's say the reward cycle boundary is Bitcoin Block 10, and our `burnchain_tip` is 9. If we receive unsolicited block data with a consensus hash pointing at block 10, our `burnchain_tip`'s active reward set will be different than the unsolicited block's reward set.

Reply:
Yes, because it's often the case that you'll receive a Stacks block for a Bitcoin block you haven't yet processed.  In all but the first-ever Bitcoin block in that reward cycle, you already know the reward set to use to authenticate the block (so the node should do so in order to buffer up the soon-to-be-processable block).

Reply:
That doesn't seem right -- this won't _error_ if the block's corresponding reward set isn't processed yet, it will just use the wrong active reward set. So what happens when it receives unsolicited blocks from the next reward cycle (because the Sortition DB hasn't caught up yet), does it buffer them anyways (seems like it would do if the reward set was a close mutation)? Does it reject the block, banning that peer as misbehaving? Does it just drop the block?

Reply:
So, the fundamental problem here is that there's no way to pair an unsolicited Nakamoto block to a sortition if its consensus hash is unrecognized.  This, in turn, makes it impossible to determine if the Nakamoto block's consensus hash or signatures are valid.  This could be due to one of two reasons:

* The consensus hash is invalid (e.g. doesn't exist in any sortition history), in which case, Nakamoto block should not be buffered since it can never be processed

* The consensus hash is valid, but the node doesn't know it yet (e.g. it's in an upcoming sortition).  In this case, the node should buffer the block because it will soon be processable.

Unfortunately, there's no way to distinguish between these two reasons.  The code here will buffer up the block if it happens to be signed by a threshold of the current reward set's signers, but the buffered block (along with all other buffered messages) will be forwarded to the relayer and dropped from `self.pending_messages` when the network stack's view of the burnchain (`self.burnchain_tip`) changes.  The rationale is that there's no real harm done by buffering up blocks that cannot yet be validated against a snapshot, because at worst they'll be dropped by the relayer.

To answer your specific questions:

> So what happens when it receives unsolicited blocks from the next reward cycle (because the Sortition DB hasn't caught up yet), does it buffer them anyways (seems like it would do if the reward set was a close mutation)? 

Yes, as long as the block signature can be validated with the _current_ reward set (which, if the signer set is stable, will be likely).  Per the above, there's no real downside of doing this, because (1) the number of buffered blocks is finite, and (2) once the node's burnchain view changes, the buffered messages will be forwarded to the relayer for processing and dropped from `PeerNetwork`.  The relayer will accept or reject them, but they'll stop taking up memory either way.

> Does it reject the block, banning that peer as misbehaving?

Because it's impossible to know at this point whether or not the block is valid, it can't be treated as misbehavior.  The peer will not be punished if it sends a Nakamoto block that turns out to be invalid.  However, it is very difficult to construct such a Nakamoto block in the first place, because 70% of the signers in the current reward cycle would need to have signed off on it.

> Does it just drop the block?

It tries to hold onto it, and will forward it to the relayer to handle it once the burnchain view advances.

Reply:
Right -- that answers the first direction of misbehavior: accepting a block that is signed correctly by Set N, but not correctly by N+1, and that all makes sense. I'm also asking about the other direction: a block is correctly signed by Set N+1, but my node considers it using Set N, which it fails. What happens in that situation? Does the (incorrectly) rejected block get dropped? Is the peer that sent the (incorrectly) rejected block punished?

Reply:
> a block is correctly signed by Set N+1, but my node considers it using Set N, which it fails. What happens in that situation? Does the (incorrectly) rejected block get dropped? Is the peer that sent the (incorrectly) rejected block punished?

Assuming that set N is sufficiently different from set N+1 that signature verification fails, then the block is simply dropped with no consequence to the peer.  It's considered unbufferable.  That's the way it is right now in this PR.

The system currently enforces a maximum per-peer per-pushed-message bandwidth limit.  Thanks to your prompting, I've noticed that this PR does not add one for pushed Nakamoto blocks.  It definitely should, so someone can't just hammer the node with bad blocks and force it to consume CPU and RAM validating pushed blocks.  In addition, the checks in `buffer_data_message()` should be carried out _before_ validation.  I'll fix both problems.

Reply:
Okay, this has been addressed.  While it continues to be the case that a block with a signature verification failure will simply be dropped with no consequence to the sender, a malicious sender cannot overwhelm the node or force it to do any extra work to validate the block once its Nakamoto block message buffer is full and/or once the node's Nakamoto block message bandwidth limit is exceeded.

Furthermore, a block will only be buffered if the signature verification succeeds and if the block has not already been processed, so it's not like the malicious sender can flood the victim with garbage data.  Each buffered block will have been signed by a valid PoX signing set.